### ISSUE TITLE
when custom the objective function in xgboost, what should be the shape of the hessian?

### ISSUE THREAD

**Issue by Cai-SunShine** [dmlc/xgboost#10259]
Hi, recently I works in customizing the objective function in xgboost. I also have a doubt about the `multi:softprob` in [tutorial](https://xgboost.readthedocs.io/en/stable/tutorials/custom_metric_obj.html). In the tutorial, the hessian is given in the shape of (n_samples, n_classes). However, when do Taylor expansion for a multi variable function, the hessian should be a matrix, but not a vector. I think the correct for the hessian should be the shape of (n_samples, n_classes, n_classes). 
This is my derivation process for solving the loss function for classification tasks.

```math
\mathcal{L}^{(t)}=\sum_{i=1}^n \mathcal{l}(y_i,\hat{y}_i^{(t-1)}+f_t(x_i))+\Omega(f_t) 
```

```math
\widetilde{-}\sum_{i=1}^n [\mathcal{l}(y_i,\hat{y}_i) +f_t(x_i)^Tg_i+f_t(x_i)^Th_if_t(x_i)]+\gamma T + \frac{1}{2}\lambda \sum_{j=1}^T||w_j||_2 
```

```math
\widetilde{-}\sum_{i=1}^n [f_t(x_i)^Tg_i+f_t(x_i)^Th_if_t(x_i)] + \gamma T + \frac{1}{2}\lambda \sum_{j=1}^T||w_j||_2\\
```

```math
\widetilde{-}\sum_{j=1}^T [(\sum_{i\in I_j}g^T_i)w_j + \frac{1}{2}w_j^T(\sum_{i\in I_j}h_i + \lambda I)w_j] + \gamma T
```

the question can be concluded as solve the optimization problem about
```math
\min \sum_{j=1}^T [(\sum_{i\in I_j}g^T_i)w_j + \frac{1}{2}w_j^T(\sum_{i\in I_j}h_i + \lambda I)w_j] + \gamma T

```

the $g_i$ I think is the matrix shape but not vector shape.

Thanks for your time for reading my post!

**Comment by hcho3** [comment]
We use the diagonal approximation, so non-diagonal entries in the Hessian matrix are discarded.

**Comment by Cai-SunShine** [comment]
> We use the diagonal approximation, so non-diagonal entries in the Hessian matrix are discarded.

Thanks for reply. So it is not the precious formula of second order Taylor, just a approximation? But if drop all the non-diagonal, may be it brings some errors? Have you mentioned it in the [xgboost paper](https://arxiv.org/abs/1603.02754)

**Comment by hcho3** [comment]
In practice, diagonal approximation brings great performance benefit (less memory consumption, less computation) while still achieving similar accuracy.

**Comment by Cai-SunShine** [comment]
> In practice, diagonal approximation brings great performance benefit (less memory consumption, less computation) while still achieving similar accuracy.

Thanks for your reply!
